[libc++][ranges] Implement ranges::contains_subrange

ZijunZhaoCCK opened the pull request.

ZijunZhaoCCK committed:
[libc++] Implement ranges::contains_subrange

ZijunZhaoCCK committed:
format

ZijunZhaoCCK committed:
Add a newline

huixie90 commented on libcxx/include/__algorithm/ranges_contains_subrange.h:
```diff
@@ -0,0 +1,145 @@
+      _Proj1& __proj1,
+      _Proj2& __proj2,
+      _Offset __offset) {
+    if (__offset < 0)
+      return false;
+    else {
+      for (; __offset >= 0; __offset--, __first1++) {
+        auto result = ranges::starts_with(
+            std::move(__first1),
+            std::move(__last1),
+            std::move(__first2),
+            std::move(__last2),
+            std::ref(__pred),
+            std::ref(__proj1),
+            std::ref(__proj2));
+        if (result)
+          return true;
+      }
+      return false;
+    }
```
is there a reason why you did not define in terms of `ranges::search` as in the standard?

ZijunZhaoCCK replied:
Yes.
Initially, I didn't aware that I can use `ranges::search`.
Then I tried to replace `ranges::starts_with` with `ranges::search` and I got 

> constraints not satisfied for class template 'subrange' [with _Iter = cpp20_input_iterator<int *>, _Sent = cpp20_input_iterator<int *>, _Kind = std::ranges::subrange_kind::unsized]

error in the test file.

I tried to test with forward iterator. It works well.

huixie90 replied:
`contains_subrange` only worked for `forward_iterator`, which is the same as `ranges::search`. Why do you test it with an `input_iterator`? It will fail as expected.

huixie90 commented on libcxx/include/__algorithm/ranges_contains_subrange.h:
```diff
@@ -0,0 +1,145 @@
+      _Pred __pred   = {},
+      _Proj1 __proj1 = {},
+      _Proj2 __proj2 = {}) const {
+    auto __n1     = ranges::distance(__first1, __last1);
+    auto __n2     = ranges::distance(__first2, __last2);
+    auto __offset = __n1 - __n2;
```
I would avoid using auto when dealing with difference_type, as here __n1 and __n2 are actually different types (iter_difference_t<Iter1> and iter_difference_t`). and `__offset` will be the type that has the integer ( or integer class type) with greater width  among the two

huixie90 commented on libcxx/include/__algorithm/ranges_contains_subrange.h:
```diff
@@ -0,0 +1,145 @@
+    requires indirectly_comparable<iterator_t<_Range1>, iterator_t<_Range2>, _Pred, _Proj1, _Proj2>
+  _LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI constexpr bool operator()(
+      _Range1&& __range1, _Range2&& __range2, _Pred __pred = {}, _Proj1 __proj1 = {}, _Proj2 __proj2 = {}) const {
+    auto __n1 = 0;
+    auto __n2 = 0;
```
`__n1` and `__n2` are `int`s, and later you assigned to `iter_difference_t`. This could be a hard error because integer class is not guaranteed to be "explicitly" convertible to `int`

huixie90 commented on libcxx/include/__algorithm/ranges_contains_subrange.h:
```diff
@@ -0,0 +1,145 @@
+    auto __n1 = 0;
+    auto __n2 = 0;
+
+    if constexpr (sized_range<_Range1> && sized_range<_Range2>) {
+      __n1 = ranges::size(__range1);
+      __n2 = ranges::size(__range2);
+    } else {
+      __n1 = ranges::distance(ranges::begin(__range1), ranges::end(__range1));
+      __n2 = ranges::distance(ranges::begin(__range2), ranges::end(__range2));
+    }
+
```
`ranges::distance(r)` does exactly this

huixie90 reviewed (changes requested):
Thank you for contributing! I have not looked at the tests yet. Please let me know if my comments on the header file is unclear. Thank you!

ZijunZhaoCCK committed:
1. call distance(r) directly 2. replace auto with int to avoid potential errors

huixie90 commented on libcxx/include/__algorithm/ranges_contains_subrange.h:
```diff
@@ -0,0 +1,145 @@
+      _Pred __pred   = {},
+      _Proj1 __proj1 = {},
+      _Proj2 __proj2 = {}) const {
+    int __n1     = ranges::distance(__first1, __last1);
+    int __n2     = ranges::distance(__first2, __last2);
+    int __offset = __n1 - __n2;
```
This still does not work.  `iter_difference_t` could be any integer class type, which is not guaranteed  to be explicitly convertible to `int`. Even it does, you don't want to do it, because this might be a narrowing conversion and you just get into undefined behaviour if actual distance is larger than the max of `int`

huixie90 commented on libcxx/include/__algorithm/ranges_contains_subrange.h:
```diff
@@ -0,0 +1,145 @@
+    requires indirectly_comparable<iterator_t<_Range1>, iterator_t<_Range2>, _Pred, _Proj1, _Proj2>
+  _LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI constexpr bool operator()(
+      _Range1&& __range1, _Range2&& __range2, _Pred __pred = {}, _Proj1 __proj1 = {}, _Proj2 __proj2 = {}) const {
+    int __n1 = 0;
+    int __n2 = 0;
+
+    if constexpr (sized_range<_Range1> && sized_range<_Range2>) {
+      __n1 = ranges::size(__range1);
+      __n2 = ranges::size(__range2);
+    } else {
+      __n1 = ranges::distance(__range1);
+      __n2 = ranges::distance(__range2);
+    }
```
as in the other comments. `int` won't work here

huixie90 commented on libcxx/include/__algorithm/ranges_contains_subrange.h:
```diff
@@ -0,0 +1,145 @@
+        std::move(__offset));
+  }
+
+  template <input_range _Range1,
+            input_range _Range2,
```
The constraints are wrong.  this is what is in the standard
https://eel.is/c++draft/alg.contains

ZijunZhaoCCK committed:
Reformat

shafik commented:
Please make sure you add a description to your PR. This is what usually goes into the git log and we want those entries to be as descriptive and helpful for folks who read the git logs, thank you.

philnik777 commented:
> Please make sure you add a description to your PR. This is what usually goes into the git log and we want those entries to be as descriptive and helpful for folks who read the git logs, thank you.

Do you have any specific suggestions what should be in the message? The libc++ "I implement something" patches are most of the time quite straight-forward, since you can look at the standard and know what's going on. It's also the best source, since it's the authoritative document here, and is also what anybody who cares to actually look at the patch should check against for correctness.

ZijunZhaoCCK committed:
Correct input type and call ranges::search()

var-const commented:
> > Please make sure you add a description to your PR. This is what usually goes into the git log and we want those entries to be as descriptive and helpful for folks who read the git logs, thank you.
> 
> Do you have any specific suggestions what should be in the message? The libc++ "I implement something" patches are most of the time quite straight-forward, since you can look at the standard and know what's going on. It's also the best source, since it's the authoritative document here, and is also what anybody who cares to actually look at the patch should check against for correctness.

+1 -- I'm in favor of descriptive commit messages in general, but in this case I'm not sure what would be useful to add beyond "Implement ranges::contains_subrange".

var-const commented on libcxx/include/__algorithm/ranges_contains_subrange.h:
```diff
@@ -0,0 +1,145 @@
+    int __n1 = 0;
+    int __n2 = 0;
+
+    if constexpr (sized_range<_Range1> && sized_range<_Range2>) {
```
I think we can just rely on `ranges::search` to do these checks for us, which would significantly simplify the implementation. I'd suggest just doing the check for whether the subrange is empty (it's important because `search` would return `false` in that case while we want to return `true`), then simply delegating to `search`. Optionally, you might also remove `__contains_subrange_fn_impl` (it might no longer pull its weight).

var-const reviewed (changes requested):
Thank you for the patch!

ldionne commented:
Gentle ping. There's outstanding feedback to address on this review. @ZijunZhaoCCK if you don't think you'll have time to pursue this PR anymore, it's all good but please let us know so we can assign it to someone else!

ZijunZhaoCCK commented:
> Gentle ping. There's outstanding feedback to address on this review. @ZijunZhaoCCK if you don't think you'll have time to pursue this PR anymore, it's all good but please let us know so we can assign it to someone else!

Sorry for the late reply! I can’t make it a priority right now, but I’ll definitely do it when things calm down, maybe in January. If someone want to do it right now, I can stop pursuing this PR.

mordante commented on libcxx/include/__algorithm/ranges_contains_subrange.h:
```diff
@@ -0,0 +1,145 @@
+            class _Proj1 = identity,
+            class _Proj2 = identity>
+    requires indirectly_comparable<_Iter1, _Iter2, _Pred, _Proj1, _Proj2>
+  _LIBCPP_NODISCARD_EXT _LIBCPP_HIDE_FROM_ABI constexpr bool operator()(
```
The `_LIBCPP_NODISCARD_EXT` need to be tested. Can you add them too
`libcxx/test/libcxx/diagnostics/ranges.nodiscard_extensions.compile.pass.cpp`
`libcxx/test/libcxx/diagnostics/ranges.nodiscard_extensions.verify.cpp`

mordante commented on libcxx/include/__algorithm/ranges_contains_subrange.h:
```diff
@@ -0,0 +1,145 @@
+} // namespace __contains_subrange
+
+inline namespace __cpo {
+inline constexpr auto contains_subrange = __contains_subrange::__fn{};
```
This function needs to be exported from modules. See https://github.com/llvm/llvm-project/blob/main/libcxx/modules/std/algorithm.inc#L44

mordante reviewed (commented):
Thanks for your patch!
I mainly glossed over the patch, and left some comments.

EricWF commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.contains/ranges.contains_subrange.pass.cpp:
```diff
@@ -0,0 +1,303 @@
+          subrange.begin(),
+          subrange.end(),
+          {},
+          [](int i) { return i - 3; },
```
Repetition is fine if not preferred in tests a lot of the time.

var-const replied:
Which is why I marked my comment as optional, even though I personally slightly prefer avoiding repetition in cases where it's easy to do so.

philnik777 commented on libcxx/include/__algorithm/ranges_contains_subrange.h:
```diff
@@ -0,0 +1,145 @@
+          std::ref(__pred),
+          std::ref(__proj1),
+          std::ref(__proj2));
+      return result.empty() == false;
```
```suggestion
      return !result.empty();
```

philnik777 commented on libcxx/include/__algorithm/ranges_contains_subrange.h:
```diff
@@ -0,0 +1,145 @@
+      _Sent2 __last2,
+      _Pred __pred   = {},
+      _Proj1 __proj1 = {},
+      _Proj2 __proj2 = {}) const {
```
Same here: let's make these static member functions.

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
1. update function 2. add nondiscard tests

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK commented on libcxx/test/std/algorithms/alg.nonmodifying/alg.contains/ranges.contains_subrange.pass.cpp:
```diff
@@ -0,0 +1,303 @@
+#include <vector>
+
+#include "almost_satisfies_types.h"
+#include "boolean_testable.h"
```
no, removed

ZijunZhaoCCK committed:
1. Call ranges::search() in ranges::contains_subrange()
2. add several test cases: like when both range and subrange are empty
3. reformat
4. nit to fix
5. add ranges::contains_subrange() in module.modulemap.in

ZijunZhaoCCK committed:
update code

ZijunZhaoCCK committed:
1. Call ranges::search() in ranges::contains_subrange()
2. add several test cases: like when both range and subrange are empty
3. reformat
4. nit to fix
5. add ranges::contains_subrange() in module.modulemap.in

ZijunZhaoCCK committed:
1. Call ranges::search() in ranges::contains_subrange()
2. add several test cases: like when both range and subrange are empty
3. reformat
4. nit to fix

ZijunZhaoCCK committed:
update code

ZijunZhaoCCK committed:
1. Call ranges::search() in ranges::contains_subrange()
2. add several test cases: like when both range and subrange are empty
3. reformat
4. nit to fix
5. add ranges::contains_subrange() in module.modulemap.in

ZijunZhaoCCK committed:
1. Call ranges::search() in ranges::contains_subrange()
2. add several test cases: like when both range and subrange are empty
3. reformat
4. nit to fix
5. add ranges::contains_subrange() in module.modulemap.in

ZijunZhaoCCK committed:
1. Call ranges::search() in ranges::contains_subrange()
2. add several test cases: like when both range and subrange are empty
3. reformat
4. nit to fix
5. add ranges::contains_subrange() in module.modulemap.in

ZijunZhaoCCK committed:
update code

ZijunZhaoCCK committed:
1. Call ranges::search() in ranges::contains_subrange()
2. add several test cases: like when both range and subrange are empty
3. reformat
4. nit to fix
5. add ranges::contains_subrange() in module.modulemap.in

ZijunZhaoCCK committed:
1. Call ranges::search() in ranges::contains_subrange()
2. add several test cases: like when both range and subrange are empty
3. reformat
4. nit to fix
5. add ranges::contains_subrange() in module.modulemap.in

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

H-G-Hristov commented:
Just a gentle reminder. You should probably also update the release notes, the status page and the feature test macro: `#define __cpp_lib_ranges_contains 20XXXXL // also in <algorithm>`

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

var-const commented on libcxx/docs/Status/RangesAlgorithms.csv:
```diff
@@ -3,13 +3,13 @@ C++20,all C++20 algorithms,N/A,N/A,✅
 C++23,`iota (algorithm) <https://wg21.link/p2440r1>`_,Unassigned,No patch yet,Not started
 C++23,`fold <https://wg21.link/p2322r5>`_,Unassigned,No patch yet,Not started
-C++23,`contains <https://wg21.link/p2302r2>`_,Zijun Zhao,No patch yet,In Progress
+C++23,`contains <https://wg21.link/p2302r2>`_,Zijun Zhao,N/A,Complete
```
I think you should replace `N/A` with something like
```
`#65148 <https://github.com/llvm/llvm-project/pull/65148>`_
```

var-const commented on libcxx/docs/Status/RangesAlgorithms.csv:
```diff
@@ -3,13 +3,13 @@ C++20,all C++20 algorithms,N/A,N/A,✅
-C++23,`starts_with <https://wg21.link/P1659R3>`_,Zijun Zhao,`D150735 <https://llvm.org/D150735>`_,✅
-C++23,`ends_with <https://wg21.link/P1659R3>`_,Zijun Zhao,No patch yet,In Progress
+C++23,`starts_with <https://wg21.link/P1659R3>`_,Zijun Zhao,`D150735 <https://llvm.org/D150735>`_,Complete
+C++23,`ends_with <https://wg21.link/P1659R3>`_,Zijun Zhao,N/A,Complete
```
Instead of `N/A`:
```
`D150831 <https://llvm.org/D150831>`_
```

var-const reviewed (approved):
LGTM. Thanks a lot for pushing this through and addressing all the feedback!

var-const commented:
Looks like the CI failure was fixed by #77672 -- merging `main` should resolve this.

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange

ZijunZhaoCCK committed:
Merge branch 'llvm:main' into libcxx-ranges-contains-subrange
